src/chains/ethereum/ethereum/src/api.ts

```diff
@@ -2880,6 +2880,186 @@ export default class EthereumApi implements Api {
+   * the transactions will be sorted in ascending order by effective tip per gas and the corresponding effective tip for the percentile
+   * will be determined, accounting for gas consumed.
+   * @returns transaction base fee per gas and effective priority fee per gas for the requested/supported block range
+   *
+   * * `oldestBlock`:  - Lowest number block of the returned range.
+   * * `baseFeePerGas`:  - An array of block base fees per gas. This includes the next block after the newest of the returned range,
+   * because this value can be derived from the newest block. Zeroes are returned for pre-EIP-1559 blocks.
+   * * `gasUsedRatio`:  - An array of block gas used ratios. These are calculated as the ratio of `gasUsed` and `gasLimit`.
+   * * `reward`:  - (optional) An array of effective priority fee per gas data points from a single block. All zeroes are returned if the
+   * block is empty.
+   *
+   * @example
+   * ```javascript
+   * const feeHistory = await provider.request({ method: "eth_feeHistory", params: ["0x5", "0xE5A51D", [10, 50, 80]] });
+   * console.log(feeHistory);
+   * ```
+   */
+  @assertArgLength(3)
+  async eth_feeHistory(
+    blockCount: QUANTITY,
+    newestBlock: QUANTITY | Ethereum.Tag,
+    rewardPercentiles: number[]
+  ): Promise<Ethereum.FeeHistory<"private">> {
+    const blockchain = this.#blockchain;
+    const PAD_PRECISION = 16;
+    const PRECISION_FLOAT = 1e14;
+    const PRECISION_BIG_INT = BigInt(1e16);
+
+    const newestBlockNumber = blockchain.blocks
+      .getEffectiveNumber(newestBlock)
+      .toNumber();
+
+    // blockCount > newestBlock is technically valid but we cannot go past the Genesis Block.
+    const totalBlocks = Math.min(
+      Quantity.toNumber(blockCount),
+      newestBlockNumber + 1
+    );
+
+    // Cut out early if no range is given.
+    if (totalBlocks === 0) {
+      return {
+        oldestBlock: Quantity.from(newestBlockNumber),
+        baseFeePerGas: undefined,
+        gasUsedRatio: null,
+        reward: undefined
+      };
+    }
+
+    const baseFeePerGas = new Array(totalBlocks);
+    const gasUsedRatio = new Array(totalBlocks);
+    let reward;
+    if (rewardPercentiles.length > 0) {
+      reward = new Array(totalBlocks);
+    }
+    // blockCount is inclusive of newestBlock
+    const oldestBlockNumber = newestBlockNumber - (totalBlocks - 1);
+    let currentBlockNumber = oldestBlockNumber;
+    let currentBlock: Block;
+
+    while (currentBlockNumber <= newestBlockNumber) {
+      currentBlock = await blockchain.blocks.get(
+        Quantity.toBuffer(currentBlockNumber)
+      );
+
+      const currentPosition = currentBlockNumber - oldestBlockNumber;
+
+      baseFeePerGas[currentPosition] =
+        currentBlock.header.baseFeePerGas || Quantity.Zero;
+
+      const { gasUsed, gasLimit } = currentBlock.header;
+      if (gasUsed === gasLimit) {
```

Comment:
This isn't testing _value_ equality, but reference equality, since these are both `Quantity`. 

You should probably be doing `gasUsed.toBigInt() === gasLimit.toBigInt()` here.

That said, it is probably very rare for `gasUsed` to be equal to `gasLimit`, right? If so, do you think this optimization is beneficial?

Reply:
So this was something I didn't like - the `Number(``0.${}`) monster. @MicaiahReid correctly called out the bug (When `gasUsed === gasLimit, 0.1`

Here are some of the things I tried, and the replacement. It does not behave the way I would expect. 

Assumption: `Number.isSafeInteger(a) && Number.isSafeInteger(b) && Number.isSafeInteger(c)` for any math ops involving a,b, and c or precision is lost.

Each of these was tested against:
1. `--miner.blockGasLimit=21000 --miner.defaultTransactionGasLimit=21000` (send 1 tx, `gasUsedRatio = 1`)
2. `--miner.defaultTransactionGasLimit=21000` (send 1 tx, `gasUsedRatio = .0007`)
3. Against Infura for block `0xE5A51D` (`gasUsedRatio = 0.9997199997386664`)


```javascript
const PRECISION_FLOAT: number = 1e14;
const PRECISION_BIG_INT: bigint = BigInt(1e16);
```

``` javascript
// Case #1
gasUsedRatio[currentPosition] =
        Number((gasUsed.toBigInt() * PRECISION_BIG_INT) / gasLimit.toBigInt()) /
        Number(PRECISION_BIG_INT);
```

ER: One `Number` conversion could be > `MAX_SAFE_INTEGER` and the other is, I expected Math precision to be off by `.000000000001`. 
AR: Math is correct

``` javascript
// Case #2
gasUsedRatio[currentPosition] =
        Number((gasUsed.toBigInt() * PRECISION_BIG_INT) / gasLimit.toBigInt()) /
        PRECISION_FLOAT / 100;
```

ER: The first `Number` conversion could be > `MAX_SAFE_INTEGER` (but is not with my settings), divide by `PRECISION_FLOAT` to get the percentile. Divide by `100` to get the ratio. 
AR: Test 1 pass, Test 2 fails, `0.0007000000000000001`, Test 3 pass


``` javascript
// Case #3
gasUsedRatio[currentPosition] =
        Number((gasUsed.toBigInt() * PRECISION_BIG_INT) / gasLimit.toBigInt()) /
        1e16;
```

ER: 1e16 is blatantly > `MAX_SAFE_INTEGER`  
AR: Math is correct

This is kind of the opposite of what I would expect.

Reply:
:+1: